### PR TITLE
MBS-10166: Add genre entity to the WS

### PR DESCRIPTION
**Pull request by reosarevok**
https://tickets.metabrainz.org/browse/MBS-10166

Goes with https://github.com/metabrainz/mmd-schema/pull/21

### CODE REVIEW

File: lib/MusicBrainz/Server/Controller/WS/2/Genre.pm [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+sub genre_search : Chained('root') PathPart('genre') Args(0) {
+    my ($self, $c) = @_;
+
+    $c->detach('genre_browse') if $c->stash->{linked};
+    $self->_search($c, 'genre');
+}
+
+__PACKAGE__->meta->make_immutable;
+1;
+
```
Comment (yvanzo) [inline lib/MusicBrainz/Server/Controller/WS/2/Genre.pm]: You may want to add our ending boilerplate from [Perl module template](https://musicbrainz.org/doc/Development/Server_Coding_Style#Perl_Module_Template_and_Documentation_Guidelines)

File: t/sql/webservice.sql [hunk 2]
```diff
@@ -1165,8 +1165,8 @@ INSERT INTO tag (id, name, ref_count) VALUES (33017, 'electronica dubstep', 14);
 -- Genres
 INSERT INTO genre (id, gid, name) VALUES (1, '1b50083b-1afa-4778-82c8-548b309af783', 'dubstep');
 INSERT INTO genre (id, gid, name) VALUES (2, '89255676-1f14-4dd8-bbad-fca839d6aff4', 'electronic');
-INSERT INTO genre (id, gid, name) VALUES (3, '51cfaac4-6696-480b-8f1b-27cfc789109c', 'grime');
-INSERT INTO genre (id, gid, name) VALUES (4, 'eba7715e-ee26-4989-8d49-9db382955419', 'j-pop');
+INSERT INTO genre (id, gid, name, comment) VALUES (3, '51cfaac4-6696-480b-8f1b-27cfc789109c', 'grime', 'stuff');
+    INSERT INTO genre (id, gid, name) VALUES (4, 'eba7715e-ee26-4989-8d49-9db382955419', 'j-pop');
```
Comment (mwiencek) [inline t/sql/webservice.sql]:
```suggestion
INSERT INTO genre (id, gid, name) VALUES (4, 'eba7715e-ee26-4989-8d49-9db382955419', 'j-pop');
```

File: lib/MusicBrainz/Server/Controller/WS/2/Genre.pm [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+    my ($resource, $id) = @{ $c->stash->{linked} };
+    my ($limit, $offset) = $self->_limit_and_offset($c);
+
+    if (!is_guid($id)) {
+        $c->stash->{error} = "Invalid mbid.";
+        $c->detach('bad_req');
+    }
+
+    my $genres;
```
Comment (mwiencek) [inline lib/MusicBrainz/Server/Controller/WS/2/Genre.pm]: This doesn't appear to be initialized anywhere, but it's being accessed/passed around twice below (even `$genres->{items}`, which Perl allows, but).

File: lib/MusicBrainz/Server/Controller/WS/2/Genre.pm [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+};
+
+sub base : Chained('root') PathPart('genre') CaptureArgs(0) { }
+
+sub genre_toplevel {
+    my ($self, $c, $stash, $genres) = @_;
+
+}
+
+sub genre_browse : Private {
```
Comment (mwiencek) [inline lib/MusicBrainz/Server/Controller/WS/2/Genre.pm]: Given the previous comment I'm guessing browsing isn't working yet (or I'm not sure how to use it). I'd expect, I guess, to be able to get genres for the Beatles via `/ws/2/genre/?artist=b10bbbfc-cf9e-42e0-be17-e2c3e1d2600d`? But that gives me `Invalid resource genre.`.

So we should either finish the implementation or remove the broken bits and leave them for a later PR (returning a 501 for now, I guess, like with search).

Comment (yvanzo) [reply lib/MusicBrainz/Server/Controller/WS/2/Genre.pm]: Yes, only lookup is implemented, neither browse nor search.

Comment (reosarevok) [reply lib/MusicBrainz/Server/Controller/WS/2/Genre.pm]: I'm not sure browse makes sense, honestly. Do you feel it should be added?

Comment (mwiencek) [reply lib/MusicBrainz/Server/Controller/WS/2/Genre.pm]: Not sure - I know you'd normally get them via just inc=genres or user-genres on the entity request, but if we were to add things like aliases and relationships, those probably wouldn't be included there, and a browse request might make sense. Though it's not something we have to include in this PR (I'd just remove the broken code for now).

Comment (reosarevok) [reply lib/MusicBrainz/Server/Controller/WS/2/Genre.pm]: I'm returning 501 for now.

File: lib/MusicBrainz/Server/Controller/WS/2/Genre.pm [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+with 'MusicBrainz::Server::Controller::WS::2::Role::Lookup' => {
+    model => 'Genre',
+};
+
+sub base : Chained('root') PathPart('genre') CaptureArgs(0) { }
+
+sub genre_toplevel {
+    my ($self, $c, $stash, $genres) = @_;
+
+}
```
[the remark this answers is not in the document]
Comment (reosarevok) [reply lib/MusicBrainz/Server/Controller/WS/2/Genre.pm]: It's needed for Lookup not to crash, but it can be fully empty, so I did that.

**Comment by reosarevok** [comment]
Docs: 
https://wiki.musicbrainz.org/index.php?title=Development%2FXML_Web_Service%2FVersion_2&type=revision&diff=73972&oldid=73521

https://wiki.musicbrainz.org/index.php?title=Development%2FJSON_Web_Service&type=revision&diff=73973&oldid=69254

**Comment by yvanzo** [comment]
Thanks for docs.

**Pull request merged**